include/boost/gil/histogram.hpp

```diff
@@ -0,0 +1,247 @@
+
+// Used to check if container is key-value type and the key is of std::integral type
+template <typename T, typename = void>
+struct has_integral_key_type : std::false_type {} ;
+
+template <typename T>
+struct has_integral_key_type
+        <
+            T,
+            typename std::enable_if<std::is_integral<typename T::key_type >::value >::type 
+        > : std::true_type {} ;
+
+} // namespace detail
+
+/// \ingroup Histogram
+/// \brief Use for indexable and resizable sequence containers 
+///
+/// This overload corresponds to resizable sequence containers which have random access capability.
+//  This would correspond to containers from STL like std::vector and std::deque. The only 
+/// difference from the next function is this function is helpful if older bin values in container
+/// is not to be erased i.e. accumulate the histograms.
+///
+template
+<
+    typename Container,
+    typename SrcView,
+    typename std::enable_if<
+                detail::is_indexable<Container>::value &&
+                detail::is_resizable<Container>::value, int>::type = 0
+>
+inline void image_histogram(SrcView const& srcview, Container &hist)
+{
+    gil_function_requires<ImageViewConcept<SrcView>>();
+    static_assert(std::is_arithmetic<typename Container::value_type>::value,
+                    "Improper container type for images.");
+    static_assert(std::is_unsigned<typename channel_type<SrcView>::type>::value,
+                    "Improper container type for signed images.");
+
+    using channel_t = typename channel_type<SrcView>::type;
+    using pixel_t = pixel<channel_t, gray_layout_t>;
+
+    hist.resize(std::numeric_limits<channel_t>::max() + 1);
+
+    for_each_pixel(color_converted_view<pixel_t>(srcview), [&hist](pixel_t const& p) {
+        ++hist[p];
+    });
+}
+
+/// \ingroup Histogram
+/// \brief Use for indexable and resizable sequence containers
+///
+/// This overload is the same as the above except that it does not produce the
+/// accumulate effect for histogram bin values. It returns a new container of
+/// size large enough to hold entire range of corresponding pixel value in the input image.
+///
+template
+<
+    typename Container,
+    typename SrcView,
+    typename std::enable_if<
+                detail::is_indexable<Container>::value &&
+                detail::is_resizable<Container>::value, int>::type = 0
+>
+inline Container image_histogram(SrcView const& srcview)
+{
+    gil_function_requires<ImageViewConcept<SrcView>>();
+    static_assert(std::is_arithmetic<typename Container::value_type>::value,
+                    "Improper container type for images.");
+    static_assert(std::is_unsigned<typename channel_type<SrcView>::type>::value,
+                    "Improper container type for signed images.");
+
+    using channel_t = typename channel_type<SrcView>::type;
+    using pixel_t = pixel<channel_t, gray_layout_t>;
+
+    Container hist(std::numeric_limits<channel_t>::max() + 1);
```

Comment:
I don't think this is a good idea! We also have pixel types like RGB64!

Reply:
Could we put a check on the channel size for specific container types? It wouldn't be clean though.

Reply:
Or have our own containers as @lpranam suggested in https://lists.boost.org/boost-gil/2020/06/0437.php